README.md
# Snackbar.js (Vanilla-Snackbar)

A simple implementation of a full width snackbar message using "vanilla js".  
View exmple at http://chrisdimoulis.com/snackbar.js

[Changelog](https://github.com/cdimoulis/snackbar.js/blob/master/changelog.md)

[![npm][npm]][npm-url]
[![node][node]][node-url]
[![deps][deps]][deps-url]
[![tests][tests]][tests-url]
[![coverage][cover]][cover-url]

## Production Dependencies
**None!** (not even jquery)

## Installation

### Node

```
npm install --save vanilla-snackbar
```

### Ruby Gem for Rails

```
gem install vanilla-snackbar
```

Or simply add to your Gemfile:
```ruby
gem 'vanilla-snackbar'
```


## Usage

### Importing/Including

**Node**

Import `vanilla-snackbar` into your module:
```js
import Snackbar from 'vanilla-snackbar';
```

**Rails**

Add to `app/assets/javascripts/application.js`. Note that you should just require snackbar, not vanilla-snackbar.
```js
//= require snackbar
```

### Style
See `src/snackbar.scss` for default style and reference if you desire to override any styles.

### Pre DOM ready
If you create a snackbar and create a message before the DOM is ready the message will be stored in a queue which which will execute once the DOM is ready.

### Create Snackbar
```javascript
  // New snackbar with defaults
  var default_snack = new Snackbar();

  // New snackbar with custom default time
  var short_snack = new Snackbar({time: 2000});

  // New snackbar where the default behviour is to manually close
  var manual_snack = new Snackbar({manual_close: true});
```

All options passed when creating the snackbar object are default. Overrides can be passed in each call to display a message.

#### Available Options
* `manual_close`: Boolean. Provide a close X button (true) vs timed close (false). *Default:* false
* `time`: ms of time before automatic close. (ignored if `manual_close: true`). *Default:* 5000
* `class`: String containing desired classes to add to snackbar. *Default:* empty

**NOTE: Default Options and Multiple Snackbar Objects**

A new Snackbar object will not inject new `#snackbar-wrapper` elements. All Snackbar objects use the same wrapper. It simply creates a new object with a different set of default options for displaying a Snackbar message. See below for overriding default options on a message specific basis as opposed to creating multiple Snackbar objects.

### Displaying Messages

Displaying a message is as simple as calling the `.message(msg, opts)` function of the Snackbar. There are also four helper methods for common Snackbar usage. All of these functions take 2 parameters:

* `msg`: the message to be displayed.
* `opts`: the options to override default options.

```javascript
  // Display a message
  snack.message('Hello World');

  // Helper functions:
  // Display a message that must be removed manually ->
  snack.stickyMessage('Acknowledge me!');
  // Display a message with a green background (adds class 'success') ->
  snack.success('You did it!');
  // Display a message with a red background (adds class 'error') ->
  snack.error("Something didn't work");
  // Display a message with a orangish/yellow background (adds class 'warn') ->
  snack.warn("I'd be careful if I were you...");
```

Creating a Snackbar message will return a Promise object. This promise object will resolve when the Snackbar has finished fading in.

### Overriding Default Options

```javascript
  // New snackbar with defaults
  var snack = new Snackbar();
  // Require user to close message just this one time
  snack.message('Read this', {manual_close: true})

  // New snackbar with custom default time
  var snack = new Snackbar({time: 2000});
  // Make this message stick longer than default
  snack.message("A slightly longer message..."), {time: 7500});
  // Add your own classes to the snackbar
  snack.message("My special snackbar", {class: 'my-snackbar your-snackbar'})
```

[npm]: https://img.shields.io/npm/v/vanilla-snackbar.svg
[npm-url]: https://npmjs.com/package/vanilla-snackbar

[node]: https://img.shields.io/node/v/vanilla-snackbar.svg
[node-url]: https://nodejs.org

[deps]: https://img.shields.io/david/cdimoulis/snackbar.js.svg
[deps-url]: https://david-dm.org/cdimoulis/snackbar.js

[tests]: https://img.shields.io/travis/cdimoulis/snackbar.js/master.svg
[tests-url]: https://travis-ci.org/cdimoulis/snackbar.js

[cover]: https://coveralls.io/repos/github/cdimoulis/snackbar.js/badge.svg?branch=master
[cover-url]: https://coveralls.io/github/cdimoulis/snackbar.js?branch=master


changelog.md
1.0.1
======

* Initial production deploy


spec/SCSSstub.js
module.exports = {}; // To replace SCSS imports


spec/snackbar-DOM.spec.js
require('../src/snackbar');

beforeAll( () => {
  clearElement(document.body);
});

afterAll( () => {
  clearElement(document.body);
});

describe('Wrapper', () => {
  // Snackbar wrapper added
  test('Snackbar _setDom wrapper added', () => {
    expect(document.getElementById('snackbar-wrapper')).toBeNull();
    let s = new Snackbar();
    expect(document.getElementById('snackbar-wrapper')).toBeInstanceOf(HTMLDivElement);
  });
});

describe('messages', () => {
  let s = new Snackbar();
  // Standard message
  test('Message', () => {
    s.message('message');
    let wrap = document.getElementById('snackbar-wrapper');
    expect(wrap.children.length).toBe(1);
    expect(wrap.lastChild.getAttribute('class')).toContain('snackbar');
    expect(wrap.lastChild.children.length).toBe(1);
    expect(wrap.lastChild.firstChild.getAttribute('class')).toContain('snackbar-text');
    expect(wrap.lastChild.firstChild.textContent).toBe('message');
  });

  // Success helper message
  test('success message', () => {
    s.success('yay');
    let wrap = document.getElementById('snackbar-wrapper');
    expect(wrap.lastChild.getAttribute('class')).toEqual(expect.stringContaining('success'));
  });

  // Error helper message
  test('error message', () => {
    s.error('oh no');
    let wrap = document.getElementById('snackbar-wrapper');
    expect(wrap.lastChild.getAttribute('class')).toEqual(expect.stringContaining('error'));
  });

  // Warn helper message
  test('warn message', () => {
    s.warn('maybe?');
    let wrap = document.getElementById('snackbar-wrapper');
    expect(wrap.lastChild.getAttribute('class')).toEqual(expect.stringContaining('warn'));
  });

  // Custom message
  test('custom message', () => {
    s.message('yay', {class: 'special'});
    let wrap = document.getElementById('snackbar-wrapper');
    expect(wrap.lastChild.getAttribute('class')).toEqual(expect.stringContaining('special'));
  });

  // Sticky helper message
  test('sticky message', () => {
    s.stickyMessage('yay');
    let wrap = document.getElementById('snackbar-wrapper');
    let el = wrap.lastChild;
    expect(el.lastChild.getAttribute('class')).toEqual(expect.stringContaining('snackbar-close'));
  });
});



function clearElement(wrapper) {
  while (wrapper.hasChildNodes()) {
    wrapper.removeChild(wrapper.lastChild);
  }
}


spec/snackbar-async.spec.js
require('../src/snackbar');

beforeEach( () => {
  clearElement(document.body);
});

// Wait for promise to finish to test the _changeOpacity function
test('_changeOpacity', () => {
  let snack = new Snackbar();
  return snack.message('wait').then( () => {
    let s = document.getElementsByClassName('snackbar')[0];
    expect(s).not.toBeUndefined();
    expect(s.style.opacity).toBe('1');
  });
});

// Wait for promise on _removeSnackbar
test('_removeSnackbar', () => {
  let snack = new Snackbar();
  snack.stickyMessage('wait');
  let s = document.getElementsByClassName('snackbar')[0];
  return snack._removeSnackbar(s).then( () => {
    expect(s.style.opacity).toBe('0');
    expect(document.getElementsByClassName('snackbar')[0]).toBeUndefined();
  });
});

// Test the onclick of a sticky message
test('onclick', () => {
  let snack = new Snackbar();
  snack.stickyMessage('click');
  let close = document.getElementsByClassName('snackbar-close')[0];
  return close.onclick().then( () => {
    expect(document.getElementsByClassName('snackbar')[0]).toBeUndefined();
  });
});

// Test the _onClose function
test('_onClose', () => {
  let snack = new Snackbar();
  // Manual close so we have to call _setClose manually
  snack.message('close', {manual_close: true});
  let s = document.getElementsByClassName('snackbar')[0];
  return snack._setClose(s, {manual_close: false, time: 500}).then( () => {
    expect(s.style.opacity).toBe('0');
    expect(document.getElementsByClassName('snackbar')[0]).toBeUndefined();
  });
});

function clearElement(wrapper) {
  while (wrapper.hasChildNodes()) {
    wrapper.removeChild(wrapper.lastChild);
  }
}


configs/webpack_bundle.config.js
const webpack = require('webpack');
const path = require('path');

// Pass in opts to setup configs
//   opts.entry: the name of the entry file
const config = {

  entry: './src/snackbar.js',
  output: {
    path: path.resolve('.', 'vendor/assets/javascripts'),
    filename: 'snackbar.min.js',
  },
  plugins: [
    new webpack.optimize.UglifyJsPlugin(),
  ],
  module: {
    rules: [{
      test: /\.js?$/,
      exclude: /(node_modules|configs)/,
      include: /(src)/,
      use: {
        loader: 'babel-loader',
        options: {
          presets: ['env']
        }
      }
    },
    {
      test: /\.scss$/,
      use:[
        'style-loader',
        'css-loader',
      {
        loader: 'sass-loader',
        options: {
          includePaths: [path.resolve('.','/src')]
        }
      }]
    }]
  }
}

module.exports = config;


src/snackbar.js
import './snackbar.scss';

/***************
Snackbar.js
https://github.com/cdimoulis/snackbar.js

A simple implementation of the snackbar message pops up at the bottom of the page.

view README.md for more in depth documentation

All options passed when creating the snackbar object are default. Overrides can
be passed in each call to display a message.

Usage:
Create a new snackbar ->
  var snack = new Snackbar();
Display a message ->
  snack.message('Hello World');

Helper functions:
Display a message that must be removed manually ->
  snack.stickyMessage('Acknowledge me!');
Display a message with a green background (adds class 'success') ->
  snack.success('You did it!');
Display a message with a red backbround (adds class 'error') ->
  snack.error("Something didn't work");
Display a message with a orangish/yellow backbround (adds class 'warn') ->
  snack.warn("I'd be careful if I were you...");

****************/

class Snackbar {
  /************
	Feature wide options. These options will be set at all time unless
  overridden by options passed in each call
    manual_close: Boolean. Provide a close X button (true) vs timed close (false)
      Default: false
    time: ms of time before automatic close. (ignored if manual_close: true)
      Default: 5000
    class: String containing desired classes.
      Default: null
  ************/
  constructor(options) {
    this.options = options || {manual_close: false, time: 5000};
    // If called before DOM is ready then maintain list to execute;
    this.pre_dom_queue = [];

    // Setup DOM
    this._setDom();
  }

  /**********
  / PUBLIC FUNCTIONS
  /**********/

  // Main function to display the snackbar
  // message: text to display
  // opt: default override options to send
  // Returns a promise that will resolve when fully faded in OR
  // null if it cannot be added yet
  message(message, opts) {
    opts = Object.assign({}, this.options, opts);
    var _snackbar = this._addSnackbar(message, opts);
    // Show if _snackbar Element is created
    if (_snackbar) {

      // Fade in and when complete set timeout for fade out if not manual close
      return _fadeIn(_snackbar).then( () => {
        this._setClose(_snackbar, opts);
      });
    }
    // Add to queue to show after DOM is ready
    else {
      this.pre_dom_queue.push({snackbar: this, message: message, opts: opts});
    }
  }

	// Helper for message that sticks until manually closed
  // message: text to display
  // opt: default override options to send
  // Returns a promise that will resolve when fully faded in OR
  // null if it cannot be added yet
  stickyMessage(message, opts) {
    opts = Object.assign({}, this.options, opts);
    opts.manual_close = true;
    return this.message(message, opts);
  }

  // Helper for success snackbar
  // message: text to display
  // opt: default override options to send
  // Returns a promise that will resolve when fully faded in OR
  // null if it cannot be added yet
  success(message, opts) {
    opts = Object.assign({}, this.options, opts);
    opts.class = opts.class || '';
    opts.class += ' success';
    return this.message(message, opts);
  }

  // Helper for error snackbar
  // message: text to display
  // opt: default override options to send
  // Returns a promise that will resolve when fully faded in OR
  // null if it cannot be added yet
  error(message, opts) {
    opts = Object.assign({}, this.options, opts);
    opts.class = opts.class || '';
    opts.class += ' error';
    return this.message(message, opts);
	}

  // Helper for warn snackbar
  // message: text to display
  // opt: default override options to send
  // Returns a promise that will resolve when fully faded in OR
  // null if it cannot be added yet
  warn(message, opts) {
    opts = Object.assign({}, this.options, opts);
    opts.class = opts.class || '';
    opts.class += ' warn';
    return this.message(message, opts);
	}

  /********
  / END PUBLIC FUNCTIONS
  *********/

  /**********
  / PRIVATE FUNCTIONS
  /**********/

  // Set the timeout for closing the snackbar if not opts.manual_close
  _setClose(snackbar, opts) {
    // If not manual_close then set timeout for removal
    return new Promise( (resolve, reject) => {
      if (!opts.manual_close) {
        setTimeout(() => {
          this._removeSnackbar(snackbar).then( () => {
            resolve();
          });
        }, opts.time);
      }
      else {
        resolve();
      }
    });
  }

  // Setup the elemends on the DOM
  _setDom() {
    var _body = document.getElementsByTagName('body')[0];
    // If the Body exists
    if (_body) {
      // Add snackbar if not in DOM already
      if (!document.getElementById('snackbar-wrapper')) {
        var _outer_wrapper = document.createElement('div');
        _outer_wrapper.id = 'snackbar-wrapper';
        _body.appendChild(_outer_wrapper);
      }
      if (this.pre_dom_queue.length > 0)
        this._flushQueue();
    }
    // if body is not available then call when DOM is ready
    else {
      _ready(() => {this._setDom();});
    }
  };

  // Flush out the pre_dom_queue
  _flushQueue() {
    for(let i = 0; i < this.pre_dom_queue.length; i++) {
      var sb = this.pre_dom_queue[i];
      sb.snackbar.message(sb.message, sb.opts);
    }
    this.pre_dom_queue = [];
  };

  // Add the snackbar
  // message: text to display
  // opt: options to send
  _addSnackbar(message, opts) {
    var _this = this;
    var _snackbar_wrapper = document.getElementById('snackbar-wrapper');
    // Only create snackbar if snackbar wrapper is in DOM
    if (_snackbar_wrapper) {
      var _snackbar = document.createElement('div');
      // Class names for snackbar element
      var snk_bar_class = 'snackbar';
      // Add option classes to snackbar
      if (opts.class) {
        snk_bar_class += ' '+opts.class;
      }
      _snackbar.className = snk_bar_class;

      // Text node
      var _text_wrapper = document.createElement('span');
      _text_wrapper.className = 'snackbar-text';
      _text_wrapper.appendChild(document.createTextNode(message));
      _snackbar.appendChild(_text_wrapper);
      // Add X for manual close
      if (opts.manual_close) {
        var _close = document.createElement('span');
        _close.className = 'snackbar-close';
        var _x = document.createTextNode('X');
        _close.appendChild(_x);
        // Apply click event for X
        _close.onclick = () => {
          // Returns the promise from removing snackbar
          return this._removeSnackbar(_snackbar);
        };
        _snackbar.appendChild(_close);
      }
      _snackbar_wrapper.appendChild(_snackbar);
      return _snackbar;
    }
  };

  // Remove a snackbar
  _removeSnackbar(_el) {
    // _fadeOut returns a promise to use for completion
    return _fadeOut(_el).then( () => {
      _el.remove();
    });
  };
  /**********
  / END PRIVATE FUNCTIONS
  /**********/
};

/*******
* Helper FUNCTIONS
*******/

// Fade in individual snackbar
// Returns a promise to use for completion
function _fadeIn(_el) {
  // Change opacity returns a promise so we are passing that up
  return _changeOpacity(_el, 1, 500);
};

// Fade out individual snackbar
// Returns a promise to use for completion
function _fadeOut(_el) {
  // Change opacity returns a promise so we are passing that up
  return _changeOpacity(_el, 0, 500);
};

// Change opacity
//   _el: element
//   value: the opacity value
//   time: the amount of time
// Returns a promise to use for completion
function _changeOpacity(_el, value, time) {
  // rate of change
  var fps = 24;
  var time_per_frame = time/fps;
  // current opacity
  var current_opacity = parseFloat(_el.style.opacity) || 0;
  // change for opacity
  var diff = value - current_opacity;
  var delta = diff/time_per_frame;

  // Return a promise so we know when this is done
  return new Promise( (resolve, reject) => {
    var interval = setInterval(change, time_per_frame);
    function change() {
      // Set new opacity
      current_opacity += delta;
      current_opacity = current_opacity < 0 ? 0 : current_opacity;
      current_opacity = current_opacity > 1 ? 1 : current_opacity;
      _el.style.opacity = current_opacity;
      // Check if done
      if (current_opacity === 1 || current_opacity === 0){
        // End interval and resolve the promise
        clearInterval(interval);
        resolve();
      }
    }
  });
};

// Callback when DOM is ready
function _ready(cb) {
  // If add event listener is available
  if (document.addEventListener) {
    document.addEventListener('DOMContentLoaded', function() {
      document.removeEventListener('DOMContentLoaded', this.callee);
      cb();
    });
  }
  // Otherwise attach the state change event
  else if (document.attachEvent) {
    document.attachEvent('onreadystatechange', function() {
      if (document.readyState === 'interactive' || document.readyState === 'complete') {
        document.detachEvent('onreadystatechange', this.callee);
        cb();
      }
    });
  }
};

window.Snackbar = Snackbar;

export default Snackbar
